### ISSUE TITLE
Fix Z3 API usage

### ISSUE THREAD

**Issue by rafaelsamenezes** [esbmc/esbmc#874]
Following the issue: https://github.com/Z3Prover/z3/issues/6479#issuecomment-1337364769

Running ESBMC with the following program:

```c
extern void abort(void);
#include <assert.h>
void reach_error() { assert(0); }

#include <stdlib.h>

int *a, *b;
int n;

extern int __VERIFIER_nondet_int(void);

void foo()
{
  int i;
  for (i = 0; i < n; i++)
    a[i] = n;
  for (i = 0; i < n - 1; i++)
    b[i] = n;
}

int main()
{
  n = 1;

  while(__VERIFIER_nondet_int() && n < 30) {
    n++;
  }

  a = malloc (n * sizeof(*a));
  b = malloc (n * sizeof(*b));

  *b++ = n;
  foo ();

  if (b[-1] - b[n - 2])
  { free(a); free(b); } /* invalid, but branch not accessible */
  else
  { free(a); free(b-1); }
  return 0;
}
```

Using `esbmc test.c --incremental-bmc --force-malloc-success --k-step 7  --z3` causes the issue.

**Comment by mikhailramalho** [comment]
It's weird because I removed our wrapper a long time ago, maybe it's the
tuple code in the z3 backend?

**Comment by rafaelsamenezes** [comment]
Not sure, but that's the only place where I see z3 vectors.

```cpp
//....
z3::expr_vector args(z3_ctx);
  for(std::size_t j = 0; j < num_fields; j++)
  {
    if(i == j)
    {
      /* use new_val at position i */
      args.push_back(newval);
    }
    else
    {
      /* use field j of t */
      z3::func_decl proj_decl =
        z3::to_func_decl(z3_ctx, Z3_get_tuple_sort_field_decl(z3_ctx, ty, j));
      args.push_back(proj_decl(t));
    }
  }

  return z3::to_func_decl(z3_ctx, Z3_get_tuple_sort_mk_decl(z3_ctx, ty))(args);
```

**Comment by mikhailramalho** [comment]
Yeah, we use the C API directly `Z3_get_tuple_sort_field_decl`, `Z3_get_tuple_sort_mk_decl`. This code probably needs to be updated to use the C++ API instead

**Comment by rafaelsamenezes** [comment]
Apparently the issue was not API related and due to how we instantiated the solver. https://github.com/esbmc/esbmc/commit/b6693a55d1efab11cc5c8274c81ba27e38d9b799 solves the issue.

@mikhailramalho Do you know if `p.set("relevancy", 0U);` is still relevant?

**Comment by mikhailramalho** [comment]
Disabling relevancy propagation increased the performance considerably. I
suggest testing this change in sv-comp.

Also, we set the print mode to be smt2 compliant or z3 would print some
weird stuff. Not sure this is still the case.

**Comment by mikhailramalho** [comment]
But enabling relevancy back changes the generated smt? Does Nikolas comment
no longer apply after this change?

Also, can't we update to use the c++ api only in that code?

**Comment by rafaelsamenezes** [comment]
> Also, we set the print mode to be smt2 compliant

The Z3 C++ API already sets it internally.

**Comment by rafaelsamenezes** [comment]
> Does Nikolas comment no longer apply after this change?

I looked at both the Z3 log file and executed with a debug build. I didn't see any call to `Z3_ast_vector_resize` while debugging. I commented at the issue but I don't think he saw it.

**Comment by rafaelsamenezes** [comment]
> Also, can't we update to use the c++ api only in that code?

Not that easy unless we stop handling tuple as sorts. I tried a hack (maintaining a cache of func_decl)... it didn't help at all.

**Comment by lucasccordeiro** [comment]
@fbrausse : can I ask whether you could look at this issue? Some messages are exchanged between Rafael and Nikolaj (please see the issue history).

**Comment by fbrausse** [comment]
I don't know what exactly the problem is here. From https://github.com/Z3Prover/z3/issues/6479#issue-1476244002:

> some performance issues

What exactly is the problem? I'm getting VERIFICATION SUCCESSFUL, after a while. [gh-874.log](https://github.com/esbmc/esbmc/files/10805236/gh-874.log)

**Comment by rafaelsamenezes** [comment]
Should be reopen with a more concrete example.
